### PR TITLE
[DO NOT MERGE]: Apply consensus and equivocation check before broadcast

### PR DESCRIPTION
**Pull request by terencechain**
Before broadcast, this PR applies consensus and equivocation checks for the beacon API's propose block endpoint.

Consensus check:
- Get pre-state from input block's parent root using stategen's state by root
- if pre state + input block to execute state transition fails, error out

Equivocation check:
- Add a cache to track seen proposer index for the block slot. The first index of the cache is the block slot
- Append proposer index after valid block signature. Clear the cache if the slot is higher than the last written slot. Do nothing if the block slot is not from the current slot

⚠️ : This is intended for relayer only. It should not be merged to `develop`

### CODE REVIEW

File: beacon-chain/node/node.go [hunk 1]
```diff
@@ -753,6 +753,11 @@ func (b *BeaconNode) registerRPCService(router *mux.Router) error {
 		return err
 	}
 
+	var regSyncService *regularsync.Service
+	if err := b.services.FetchService(&regSyncService); err != nil {
+		return err
+	}
+
```
Comment (prestonvanloon) [inline beacon-chain/node/node.go:760]: What's this used for?

File: beacon-chain/sync/service.go [hunk 2]
```diff
@@ -301,3 +302,7 @@ type Checker interface {
 	Status() error
 	Resync() error
 }
+
+type EqChecker interface {
```
Comment (rauljordan) [inline beacon-chain/sync/service.go:306]: I would call it EquivocationChecker and add a thorough comment. Eq sounds like equals in other programming languages

File: beacon-chain/sync/validate_beacon_blocks.go [hunk 3]
```diff
@@ -384,3 +384,17 @@ func getBlockFields(b interfaces.ReadOnlySignedBeaconBlock) logrus.Fields {
...
 	}
 }
+
+// HasBlock returns true if
+//  1. same block slot and proposer index is in the pending queue
+//  2. same block has been seen
+func (s *Service) HasBlock(slot primitives.Slot, proposerIdx primitives.ValidatorIndex) bool {
+	blks := s.pendingBlocksInCache(slot)
+	for _, blk := range blks {
+		if blk.Block().Slot() == slot && blk.Block().ProposerIndex() == proposerIdx {
```
Comment (potuz) [inline beacon-chain/sync/validate_beacon_blocks.go:394]: are nil checks for `blk.Block()` necessary?

File: beacon-chain/sync/validate_beacon_blocks.go [hunk 4]
```diff
@@ -384,3 +393,32 @@ func getBlockFields(b interfaces.ReadOnlySignedBeaconBlock) logrus.Fields {
...
 }
+
+func (s *Service) insertProposerIndexCache(slot primitives.Slot, proposerIdx primitives.ValidatorIndex) {
+	if s.cfg.chain.CurrentSlot() != slot {
+		return
+	}
+
+	switch {
+	case uint64(s.seenProposerIndexCache[0]) > uint64(slot):
+		return
```
Comment (potuz) [inline beacon-chain/sync/validate_beacon_blocks.go:404]: This case literally can't happen.

File: beacon-chain/sync/validate_beacon_blocks.go [hunk 5]
```diff
@@ -384,3 +393,32 @@ func getBlockFields(b interfaces.ReadOnlySignedBeaconBlock) logrus.Fields {
...
+		s.seenProposerIndexCache = append(s.seenProposerIndexCache, proposerIdx)
+	case uint64(slot) > uint64(s.seenProposerIndexCache[0]):
+		// Overwrite slot in proposer index cache if it's higher.
+		s.seenProposerIndexCache = []primitives.ValidatorIndex{primitives.ValidatorIndex(slot)}
+		s.seenProposerIndexCache = append(s.seenProposerIndexCache, proposerIdx)
+	}
+}
+
+func (s *Service) SeenProposerIndex(slot primitives.Slot, proposerIdx primitives.ValidatorIndex) bool {
+	if s.seenProposerIndexCache[0] != primitives.ValidatorIndex(slot) {
```
Comment (potuz) [inline beacon-chain/sync/validate_beacon_blocks.go:415]: nilcheck for `seenProposerIndexCache`

File: beacon-chain/sync/validate_beacon_blocks.go [hunk 6]
```diff
@@ -384,3 +393,32 @@ func getBlockFields(b interfaces.ReadOnlySignedBeaconBlock) logrus.Fields {
 		"version":       b.Block().Version(),
 	}
 }
+
+func (s *Service) insertProposerIndexCache(slot primitives.Slot, proposerIdx primitives.ValidatorIndex) {
+	if s.cfg.chain.CurrentSlot() != slot {
+		return
+	}
+
```
Comment (potuz) [inline beacon-chain/sync/validate_beacon_blocks.go:401]: nilcheck for `seenProposerIndexCache`

File: beacon-chain/sync/validate_beacon_blocks.go [hunk 7]
```diff
@@ -384,3 +393,32 @@ func getBlockFields(b interfaces.ReadOnlySignedBeaconBlock) logrus.Fields {
...
+	switch {
+	case uint64(s.seenProposerIndexCache[0]) > uint64(slot):
+		return
+	case uint64(s.seenProposerIndexCache[0]) == uint64(slot):
+		s.seenProposerIndexCache = append(s.seenProposerIndexCache, proposerIdx)
+	case uint64(slot) > uint64(s.seenProposerIndexCache[0]):
+		// Overwrite slot in proposer index cache if it's higher.
+		s.seenProposerIndexCache = []primitives.ValidatorIndex{primitives.ValidatorIndex(slot)}
+		s.seenProposerIndexCache = append(s.seenProposerIndexCache, proposerIdx)
```
Comment (potuz) [inline beacon-chain/sync/validate_beacon_blocks.go:410]:
```suggestion
		s.seenProposerIndexCache = []primitives.ValidatorIndex{primitives.ValidatorIndex(slot), proposerIdx}
```

**Review by potuz — APPROVED** [review]

File: beacon-chain/rpc/eth/beacon/blocks.go [hunk 8]
```diff
@@ -1019,6 +1018,24 @@ func (bs *Server) submitBlock(ctx context.Context, blockRoot [fieldparams.RootLe
 		})
 	}()
 
+	if block == nil || block.IsNil() {
+		return errors.New("nil block")
+	}
```
Comment (prestonvanloon) [inline beacon-chain/rpc/eth/beacon/blocks.go:1023]: Can you do this before the defer function?

**Comment by dewindtk** [comment]
I am setting up a testnet and I would insanely appreciate a Dockerfile in order to set this up in Docker. Would it be possible to get a hold of one?

**Comment by MrKoberman** [comment]
Is this PR still required when using the latest version of prysm used by the relays?

**Comment by rauljordan** [comment]
hi @dewindtk check out https://github.com/OffchainLabs/eth-pos-devnet if you want to set up a testnet. If you want to test out a specific change, such as this PR, you will need to build images with Bazel here: https://docs.prylabs.network/docs/install/install-with-bazel#running-images. We do not use dockerfiles

**Pull request closed without merging**